trio/_file_io.py

```diff
@@ -48,59 +74,199 @@
 }
 
 
-class AsyncIOWrapper(AsyncResource):
+FileT = TypeVar("FileT")
+FileT_co = TypeVar("FileT_co", covariant=True)
+T = TypeVar("T")
+T_co = TypeVar("T_co", covariant=True)
+T_contra = TypeVar("T_contra", contravariant=True)
+AnyStr_co = TypeVar("AnyStr_co", str, bytes, covariant=True)
+AnyStr_contra = TypeVar("AnyStr_contra", str, bytes, contravariant=True)
+
+# This is a little complicated. IO objects have a lot of methods, and which are available on
+# different types varies wildly. We want to match the interface of whatever file we're wrapping.
+# This pile of protocols each has one sync method/property, meaning they're going to be compatible
+# with a file class that supports that method/property. The ones parameterized with AnyStr take
+# either str or bytes depending.
+
+# The wrapper is then a generic class, where the typevar is set to the type of the sync file we're
+# wrapping. For generics, adding a type to self has a special meaning - properties/methods can be
+# conditional - it's only valid to call them if the object you're accessing them on is compatible
+# with that type hint. By using the protocols, the type checker will be checking to see if the
+# wrapped type has that method, and only allow the methods that do to be called. We can then alter
+# the signature however it needs to match runtime behaviour.
+if TYPE_CHECKING:
+    from typing_extensions import Buffer, Protocol
+
+    # fmt: off
+
+    class _HasClosed(Protocol):
+        @property
+        def closed(self) -> bool: ...
+
+    class _HasEncoding(Protocol):
+        @property
+        def encoding(self) -> str: ...
+
+    class _HasErrors(Protocol):
+        @property
+        def errors(self) -> str | None: ...
+
+    class _HasFileNo(Protocol):
+        def fileno(self) -> int: ...
+
+    class _HasIsATTY(Protocol):
+        def isatty(self) -> bool: ...
+
+    class _HasNewlines(Protocol):
+        @property
+        def newlines(self) -> Any: ...  # None, str or tuple
```

Comment:
I don't really understand this comment.

Reply:
I preserved it from the `io` type hints, looks like the docs say this returns those but typeshed chose to make it `Any`. Actually we could just use a typevar here to do whatever the file object says it returns.